### PR TITLE
feat(FEC-10980): Enable setting ABR settings on-the-fly

### PR DESCRIPTION
**Pull request by Yuvalke**
### Description of the Changes

add an API applyABRRestriction to apply ABR restrictions and show only available tracks after restriction.

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/hls-adapter.js [hunk 1]
```diff
@@ -184,15 +197,17 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
...
-
       if (abr.restrictions) {
-        if (abr.restrictions.minBitrate > 0) {
-          adapterConfig.hlsConfig.minAutoBitrate = abr.restrictions.minBitrate;
-        }
-        if (abr.restrictions.maxBitrate < Infinity) {
-          //You can either set capping by size or bitrate, if bitrate is set then disable size capping
-          adapterConfig.hlsConfig.capLevelToPlayerSize = false;
-          adapterConfig.abr.restrictions = abr.restrictions;
+        if (abr.restrictions.maxBitrate > abr.restrictions.minBitrate) {
```
Comment (yairans) [inline src/hls-adapter.js:201]: couldn't be equal for a specific one? anyway add it to the previous one

File: src/hls-adapter.js [hunk 2]
```diff
@@ -876,45 +902,44 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
...
+        if (minLevel) {
+          this._hls.minAutoBitrate = minLevel.bandwidth;
         }
       }
+      if (maxLevel) {
+        this._hls.autoLevelCapping = videoTracks.findIndex(level => level === maxLevel);
+      }
     } else {
-      this._hls.currentLevel = 0;
+      HlsAdapter._logger.warn('Doesnt meet the restriction');
```
Comment (yairans) [inline src/hls-adapter.js:942]:
```suggestion
      HlsAdapter._logger.warn('Does not meet the restriction');
```

**Review by yairans — CHANGES_REQUESTED** [review]

File: src/hls-adapter.js (hunk 1 above)
Comment (Yuvalke) [reply src/hls-adapter.js:201]: That's what we had before but it's correct

File: src/hls-adapter.js [hunk 3]
```diff
@@ -876,45 +902,44 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
...
-            this._hls.autoLevelCapping = maxLevel;
-          } else {
-            HlsAdapter._logger.warn(
-              'Invalid maxBitrate restriction, maxBitrate must be greater than minBitrate',
-              minBitrate,
-              restrictions.maxBitrate
-            );
-          }
+  _maybeApplyAbrRestrictions(restrictions: PKABRRestrictionObject): void {
+    const videoTracks = this._playerTracks.filter(track => track instanceof VideoTrack);
```
Comment (Yuvalke) [inline src/hls-adapter.js:927]: add the filtered as parameter

File: src/hls-adapter.js [hunk 4]
```diff
@@ -872,49 +888,52 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
...
-              'Invalid maxBitrate restriction, maxBitrate must be greater than minBitrate',
-              minBitrate,
-              restrictions.maxBitrate
-            );
+  _maybeApplyAbrRestrictions(restrictions: PKABRRestrictionObject): void {
+    if (this._hls.capLevelToPlayerSize) return;
+    const videoTracks = this._playerTracks.filter(track => track instanceof VideoTrack);
+    const availableTracks = filterTracksByRestriction(videoTracks, restrictions);
+    if (availableTracks.length) {
+      this._hls.capLevelToPlayerSize = false;
```
Comment (yairans) [inline src/hls-adapter.js:921]: must to already be false, else you return in the first line.

File: src/hls-adapter.js [hunk 5]
```diff
@@ -872,49 +888,52 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
...
+      const minLevel = availableTracks[0];
+      const maxLevel = availableTracks.pop();
+      if (maxLevel !== minLevel) {
+        if (minLevel) {
```
Comment (yairans) [inline src/hls-adapter.js:925]: move this if one line above

File: src/hls-adapter.js [hunk 6]
```diff
@@ -872,49 +888,52 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
...
+          if (this._hls.currentLevel < minLevel.index || this._hls.currentLevel === -1) {
+            this._hls.nextLoadLevel = minLevel.index;
```
Comment (yairans) [inline src/hls-adapter.js:927]: this should be done also when maxLevel === minLevel

**Review by yairans — CHANGES_REQUESTED** [review]

File: src/hls-adapter.js [hunk 7]
```diff
@@ -890,34 +906,29 @@ export default class HlsAdapter extends BaseMediaSourceAdapter {
...
-            this._hls.autoLevelCapping = maxLevel;
-          } else {
-            HlsAdapter._logger.warn(
-              'Invalid maxBitrate restriction, maxBitrate must be greater than minBitrate',
-              minBitrate,
-              restrictions.maxBitrate
-            );
-          }
+  _maybeApplyAbrRestrictions(restrictions: PKABRRestrictionObject): void {
+    if (this._hls.capLevelToPlayerSize) return;
```
Comment (Yuvalke) [inline src/hls-adapter.js:913]: change to multiple lines if

**Review by yairans — APPROVED** [review]

**Pull request merged**